### ISSUE TITLE
'WCS.all_world2pix' failed to converge when plotting WCS with non linear distortions

### ISSUE THREAD

**Issue by karlwessel** [astropy/astropy#11446]
### Description
When trying to plot an image with a WCS as projection that contains non linear Distortions it fails with a `NoConvergence` error.

### Expected behavior
When I add `quiet=True` as parameter to the call 
```pixel = self.all_world2pix(*world_arrays, 0)``` 
at line 326 of `astropy/wcs/wcsapi/fitswcs.py` I get the good enough looking plot below:

![bugreport](https://user-images.githubusercontent.com/64231/112940287-37c2c800-912d-11eb-8ce8-56fd284bb8e7.png)

It would be nice if there was a way of getting that plot without having to hack the library code like that.
### Actual behavior

The call to plotting the grid fails with the following error (last few lines, can provide more if necessary):

```
~/work/develop/env/lib/python3.9/site-packages/astropy/wcs/wcsapi/fitswcs.py in world_to_pixel_values(self, *world_arrays)
    324 
    325     def world_to_pixel_values(self, *world_arrays):
--> 326         pixel = self.all_world2pix(*world_arrays, 0)
    327         return pixel[0] if self.pixel_n_dim == 1 else tuple(pixel)
    328 

~/work/develop/env/lib/python3.9/site-packages/astropy/utils/decorators.py in wrapper(*args, **kwargs)
    534                     warnings.warn(message, warning_type, stacklevel=2)
    535 
--> 536             return function(*args, **kwargs)
    537 
    538         return wrapper

~/work/develop/env/lib/python3.9/site-packages/astropy/wcs/wcs.py in all_world2pix(self, tolerance, maxiter, adaptive, detect_divergence, quiet, *args, **kwargs)
   1886             raise ValueError("No basic WCS settings were created.")
   1887 
-> 1888         return self._array_converter(
   1889             lambda *args, **kwargs:
   1890             self._all_world2pix(

~/work/develop/env/lib/python3.9/site-packages/astropy/wcs/wcs.py in _array_converter(self, func, sky, ra_dec_order, *args)
   1335                     "a 1-D array for each axis, followed by an origin.")
   1336 
-> 1337             return _return_list_of_arrays(axes, origin)
   1338 
   1339         raise TypeError(

~/work/develop/env/lib/python3.9/site-packages/astropy/wcs/wcs.py in _return_list_of_arrays(axes, origin)
   1289             if ra_dec_order and sky == 'input':
   1290                 xy = self._denormalize_sky(xy)
-> 1291             output = func(xy, origin)
   1292             if ra_dec_order and sky == 'output':
   1293                 output = self._normalize_sky(output)

~/work/develop/env/lib/python3.9/site-packages/astropy/wcs/wcs.py in <lambda>(*args, **kwargs)
   1888         return self._array_converter(
   1889             lambda *args, **kwargs:
-> 1890             self._all_world2pix(
   1891                 *args, tolerance=tolerance, maxiter=maxiter,
   1892                 adaptive=adaptive, detect_divergence=detect_divergence,

~/work/develop/env/lib/python3.9/site-packages/astropy/wcs/wcs.py in _all_world2pix(self, world, origin, tolerance, maxiter, adaptive, detect_divergence, quiet)
   1869                     slow_conv=ind, divergent=None)
   1870             else:
-> 1871                 raise NoConvergence(
   1872                     "'WCS.all_world2pix' failed to "
   1873                     "converge to the requested accuracy.\n"

NoConvergence: 'WCS.all_world2pix' failed to converge to the requested accuracy.
After 20 iterations, the solution is diverging at least for one input point.
```

### Steps to Reproduce

Here is the code to reproduce the problem:
```
from astropy.wcs import WCS, Sip
import numpy as np
import matplotlib.pyplot as plt

wcs = WCS(naxis=2)
a = [[ 0.00000000e+00,  0.00000000e+00,  6.77532513e-07,
        -1.76632141e-10],
       [ 0.00000000e+00,  9.49130161e-06, -1.50614321e-07,
         0.00000000e+00],
       [ 7.37260409e-06,  2.07020239e-09,  0.00000000e+00,
         0.00000000e+00],
       [-1.20116753e-07,  0.00000000e+00,  0.00000000e+00,
         0.00000000e+00]]
b = [[ 0.00000000e+00,  0.00000000e+00,  1.34606617e-05,
        -1.41919055e-07],
       [ 0.00000000e+00,  5.85158316e-06, -1.10382462e-09,
         0.00000000e+00],
       [ 1.06306407e-05, -1.36469008e-07,  0.00000000e+00,
         0.00000000e+00],
       [ 3.27391123e-09,  0.00000000e+00,  0.00000000e+00,
         0.00000000e+00]]
crpix = [1221.87375165,  994.90917378]
ap = bp = np.zeros((4, 4))

wcs.sip = Sip(a, b, ap, bp, crpix)

plt.subplot(projection=wcs)
plt.imshow(np.zeros((1944, 2592)))
plt.grid(color='white', ls='solid')
```

### System Details

```
>>> import platform; print(platform.platform())
Linux-5.11.10-arch1-1-x86_64-with-glibc2.33
>>> import sys; print("Python", sys.version)
Python 3.9.2 (default, Feb 20 2021, 18:40:11) 
[GCC 10.2.0]
>>> import numpy; print("Numpy", numpy.__version__)
Numpy 1.20.2
>>> import astropy; print("astropy", astropy.__version__)
astropy 4.3.dev690+g7811614f8
>>> import scipy; print("Scipy", scipy.__version__)
Scipy 1.6.1
>>> import matplotlib; print("Matplotlib", matplotlib.__version__)
Matplotlib 3.3.4
```

**Comment by embray** [comment]
You could also directly call

```python
pixel = self.all_world2pix(*world_arrays, 0)
pixel = pixel[0] if self.pixel_n_dim == 1 else tuple(pixel)
```

without patching any code.  But I wonder if the WCSAPI methods shouldn't allow passing additional keyword args to the underlying WCS methods (like `all_world2pix` in this case).  @astrofrog is the one who first introduces this API I think.

**Comment by astrofrog** [comment]
I think the cleanest fix here would be that really the FITS WCS APE14 wrapper should call all_* in a way that only emits a warning not raises an exception (since by design we can't pass kwargs through). It's then easy for users to ignore the warning if they really want.

@Cadair any thoughts?

**Comment by pllim** [comment]
Is this technically a bug?

**Comment by Cadair** [comment]
> the FITS WCS APE14 wrapper should call all_* in a way that only emits a warning

This is probably the best solution. I certainly can't think of a better one.

On keyword arguments to WCSAPI, if we did allow that we would have to mandate that all implementations allowed `**kwargs` to accept and ignore all unknown kwargs so that you didn't make it implementation specific when calling the method, which is a big ugly.

**Comment by Cadair** [comment]
> Is this technically a bug?

I would say so yes.

**Comment by karlwessel** [comment]
> > the FITS WCS APE14 wrapper should call all_* in a way that only emits a warning
> 
> This is probably the best solution. I certainly can't think of a better one.
> 

That solution would be also fine for me.

**Comment by pllim** [comment]
@karlwessel , are you interested in submitting a patch for this? 😸

**Comment by karlwessel** [comment]
In principle yes, but at the moment I really can't say.

Which places would this affect? Only all calls to `all_*` in `wcsapi/fitswcs.py`?

**Comment by astrofrog** [comment]
Yes I think that's right

**Comment by mcara** [comment]
For what it is worth, my comment is about the issues with the example. I think so far, the history of `all_pix2world` shows that it is a very stable algorithm that converges for all "real" astronomical images. So, I wanted to learn about this failure. [NOTE: This does not mean that you should not catch exceptions in `pixel_to_world()` if you wish so. I have no opinion on this.]

There are several issues with the example:

**Comment by mcara** [comment]
For what it is worth, my comment is about the issues with the example. I think so far, the history of `all_pix2world` shows that it is a very stable algorithm that converges for all "real" astronomical images. So, I wanted to learn about this failure. [NOTE: This does not mean that you should not catch exceptions in `pixel_to_world()` if you wish so.]

There are several issues with the example:
1. Because `CTYPE` is not set, essentially the projection algorithm is linear, that is, intermediate physical coordinates are the world coordinates.
2. SIP standard assumes that polynomials share the same CRPIX with the WCS. Here, CRPIX of the `Wcsprm` is `[0, 0]` while the CRPIX of the SIP is set to `[1221.87375165,  994.90917378]`
3. If you run `wcs.all_pix2world(1, 1, 1)` you will get `[421.5126801, 374.13077558]` for world coordinates (and at CRPIX you will get CRVAL which is 0). This is in degrees. You can see that from the center pixel (CRPIX) to the corner of the image you are circling the celestial sphere many times.

In summary, yes `all_world2pix` can fail but it does not imply that there is a bug in it. This example simply contains large distortions (like mapping `(1, 1) -> [421, 374]`) that cannot be handled with the currently implemented algorithm but I am not sure there is another algorithm that could do better.

With regard to throwing or not an exception... that's tough. On one hand, for those who are interested in correctness of the values, it is better to know that the algorithm failed and one cannot trust returned values. For plotting, this may be an issue and one would prefer to just get, maybe, the linear approximation. My personal preference is for exceptions because they can be caught and dealt with by the caller.

**Comment by karlwessel** [comment]
The example is a minimal version of our real WCS whichs nonlinear distortion is taken from a checkerboard image and it fits it quit well:
![fitteddistortion](https://user-images.githubusercontent.com/64231/116892995-be892a00-ac30-11eb-826f-99e3635af1fa.png)

The WCS was fitted with `fit_wcs_from_points` using an artificial very small 'RA/DEC-TAN' grid so that it is almost linear.

I guess the Problem is that the camera really has a huge distortion which just isn't fitable with a polynomial. Nevertheless it still is a real camera distortion, but I agree in that it probably is not worth to be considered a bug in the `all_world2pix` method.
